sourcegen-plugin-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/gradle/builder/GradleTaskBuilder.java

```diff
@@ -0,0 +1,316 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.sourcegen.generator.visitors.gradle.builder;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.naming.NameUtils;
+import io.micronaut.inject.ast.ClassElement;
+import io.micronaut.sourcegen.annotations.GenerateGradlePlugin;
+import io.micronaut.sourcegen.annotations.GenerateGradlePlugin.Type;
+import io.micronaut.sourcegen.generator.visitors.PluginUtils;
+import io.micronaut.sourcegen.generator.visitors.PluginUtils.ParameterConfig;
+import io.micronaut.sourcegen.generator.visitors.gradle.GradlePluginUtils.GradlePluginConfig;
+import io.micronaut.sourcegen.generator.visitors.gradle.GradlePluginUtils.GradleTaskConfig;
+import io.micronaut.sourcegen.model.AnnotationDef;
+import io.micronaut.sourcegen.model.ClassDef;
+import io.micronaut.sourcegen.model.ClassDef.ClassDefBuilder;
+import io.micronaut.sourcegen.model.ClassTypeDef;
+import io.micronaut.sourcegen.model.ExpressionDef;
+import io.micronaut.sourcegen.model.FieldDef;
+import io.micronaut.sourcegen.model.InterfaceDef;
+import io.micronaut.sourcegen.model.InterfaceDef.InterfaceDefBuilder;
+import io.micronaut.sourcegen.model.MethodDef;
+import io.micronaut.sourcegen.model.MethodDef.MethodDefBuilder;
+import io.micronaut.sourcegen.model.ObjectDef;
+import io.micronaut.sourcegen.model.ParameterDef;
+import io.micronaut.sourcegen.model.StatementDef;
+import io.micronaut.sourcegen.model.TypeDef;
+import io.micronaut.sourcegen.model.VariableDef;
+
+import javax.lang.model.element.Modifier;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A builder for {@link GenerateGradlePlugin.Type#GRADLE_TASK}.
+ * Creates a task, work action and work action parameters given a plugin task configuration.
+ */
+@Internal
+public class GradleTaskBuilder implements GradleTypeBuilder {
+
+    public static final String TASK_SUFFIX = "Task";
+
+    @Override
+    public Type getType() {
+        return Type.GRADLE_TASK;
+    }
+
+    @Override
+    @NonNull
+    public List<ObjectDef> build(GradlePluginConfig pluginConfig) {
+        List<ObjectDef> objects = new ArrayList<>();
+        for (GradleTaskConfig taskConfig: pluginConfig.tasks()) {
+            objects.addAll(buildTask(pluginConfig.packageName(), taskConfig));
+        }
+        return objects;
+    }
+
+    private List<ObjectDef> buildTask(String packageName, GradleTaskConfig taskConfig) {
+        String taskType = packageName + "." + taskConfig.namePrefix() + TASK_SUFFIX;
+        ClassDefBuilder builder = ClassDef.builder(taskType)
+            .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+            .superclass(ClassTypeDef.of("org.gradle.api.DefaultTask"))
+            .addAnnotation("org.gradle.api.tasks.CacheableTask")
```

Comment:
We should probably make this configurable. It may not be worth caching, or, caching may be broken for whatever reason.